cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/SchemaValidator/index.tsx

```diff
@@ -0,0 +1,148 @@
+import cdapavsc from 'services/cdapavscwrapper';
+import {
+  SchemaGenerator,
+  generateSchemaFromComplexType,
+} from 'components/AbstractWidget/SchemaEditor/Context/SchemaGenerator';
+import { ISchemaType } from 'components/AbstractWidget/SchemaEditor/SchemaTypes';
+import { IFlattenRowType } from '../EditorTypes';
+import { isNilOrEmpty } from 'services/helpers';
+
+/**
+ * Schema validator is independent of the schema editor. It takes in a schema tree
+ * and converts to an avro schema and parses the schema through the avsc library
+ *
+ * The context provides a way to broadcast the error to appropriate row
+ *
+ * We cannot pin point the error to a specific row as avsc library doesn't
+ * provide details on the place of failure. We right now surface the error
+ * to the parent to show the error under a tree.
+ *
+ * This is separate in an effort to push this processing to a web worker in the future.
+ * If we are not able write our own avro schema parser we won't be able to pinpoint
+ * the error to a specific row. In which we have to recurrsively parse each node
+ * in the schema tree to identify specific row of issue. This kind of processing
+ * will bring the UI down which is when we will move this to the web worker.
+ */
+interface ISchemaValidatorProviderBaseState {
+  id: string;
+  time: number;
+  error: string;
+}
+interface ISchemaValidatorContext {
+  validate: (id: string, avroSchema: ISchemaType) => ISchemaValidatorProviderBaseState;
+  errorMap: Record<string, ISchemaValidatorProviderBaseState>;
+}
+const SchemaValidatorContext = React.createContext<ISchemaValidatorContext>({
+  validate: null,
+  errorMap: {},
+});
+const SchemaValidatorConsumer = SchemaValidatorContext.Consumer;
+
+class SchemaValidatorProvider extends React.Component {
+  public defaultState = {
+    errorMap: {},
+  };
+
+  /**
+   * The subtree validation takes in a flat field, its ancestors
+   * and navigates to the appropriate child in the schema tree and parses
+   * the subtree. If it is valid nothing happens. If it is invalid we add the
+   * error to the validator context with the id of the immediate parent of the field
+   *
+   * The error validation consumer is used per row and whichever row is added
+   * to the error map gets highlighted with error.
+   *
+   * The sibliling line connection gets highlighted when parent is higlighted with an
+   * error.
+   * @param field Current flat row updated by user
+   * @param schemaTree schema tree to validate against.
+   */
+  private isSubTreeValidAvroSchema = (field: IFlattenRowType, schemaTree) => {
+    const { ancestors } = field;
+    if (!ancestors || (Array.isArray(ancestors) && !ancestors.length)) {
+      return;
+    }
+    if (ancestors.length === 1) {
+      try {
+        const entireSchema = SchemaGenerator(schemaTree);
+        cdapavsc.parse(entireSchema.schema, { wrapUnions: true });
+      } catch (e) {
+        return { error: e.message, fieldIdToShowError: ancestors[0] };
```

Comment:
Do we know for sure that the error is always in.message?

Reply:
Yeah this is the standard javascript error thrown from within the library.